src/engine/guidance/collapse_scenario_detection.cpp

```diff
@@ -81,8 +95,11 @@ bool basicCollapsePreconditions(const RouteStepIterator first,
                                      hasRoundaboutType(second->maneuver.instruction) ||
                                      hasRoundaboutType(third->maneuver.instruction);
 
+    const auto contains_bad_uturn =
+        !isValidUturnCombination(first, second) && !isValidUturnCombination(second, third);
+
     // require modes to match up
-    return !has_roundabout_type && haveSameMode(*first, *second, *third);
+    return !has_roundabout_type && haveSameMode(*first, *second, *third) && !contains_bad_uturn;
```

Comment:
Technically we do this. The combination of both is what we do here. The only difference is the involvement of waypoint types, which can happen for three-hop scenarios (without changing the waypoint itself)